### PR TITLE
oath-toolkit: update to 2.6.7

### PR DESCRIPTION
**Pull request by nickgaya**
#### Description

* Update oath-toolkit to version 2.6.7
* Add workaround for missing `pam_modutil_getpwnam` function.

###### Type(s)

- [x] update

###### Tested on
macOS 10.14.6 18G9028 x86_64
Xcode 11.3.1 11C505

###### Verification
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by nickgaya** [comment]
The `pam_modutil_getpwnam` workaround allows the code to compile but it's potentially insecure, as the `getpwname(3)` function is not threadsafe.  A better alternative would be to use `getpwname_r(3)`, but this is somewhat complex to do (see the [Linux PAM implementation](https://github.com/linux-pam/linux-pam/blob/master/libpam/pam_modutil_getpwnam.c)).

@fracai Any thoughts on this?

**Comment by fracai** [comment]
The puzzle solver in me wants to figure out using getpwnam_r, but a quick glance at that link isn't clear what would need to be done. I did find another reference in a different project that refers to using _r as being extra work. I'm also not clear on the consequences of using the non-threadsafe version. Worst case that could presumably lead to improper authentication or local privilege escalation. Maybe that's doomsday thinking.
I suppose the safest, lowest effort option, might be to disable the features that require getpwnam?
Bottom line? I have more conjectures than thoughts.
I can add this to my todo list to look at further, but I can't guarantee any timeline. I appreciate your effort so far to update it.

**Comment by nickgaya** [comment]
The difficulty with `getpwnam_r` is that you need to provide a buffer for the result. Linux PAM allocates this buffer on the heap and attaches it to the PAM handle to get freed at the end of the transaction, which is kind of a hassle.

However I found that [NetBSD's pam_unix implementation](https://github.com/NetBSD/src/blob/trunk/lib/libpam/modules/pam_unix/pam_unix.c) takes the simpler approach of allocating a small fixed-size buffer on the stack. This is probably good enough for most use cases, so I'll try writing a patch to do it this way in pam_oath.

I've also been tinkering with grabbing the relevant linux-pam source files and compiling them as part of the pam_oath module (with relevant patches).  I got the port to build with this approach but it adds a fair amount of complexity and might be a headache to maintain.

**Comment by fracai** [comment]
It would also probably be worth reporting upstream. That'd be a better place to maintain it anyway.

**Comment by nickgaya** [comment]
Actually, it turns out I misread the documentation.  On Mac OS, `getpwnam` is in fact thread-safe, but not re-entrant:

> All of these routines are thread-safe.  The `getpwent()`, `getpwnam()`, and
     `getpwuid()` routines return a pointer to a result managed by the system
     library in a thread-specific data structure.  Every thread has space for
     a pointer to a `struct passwd` and allocates its own storage for the
     result.  Neither previously returned values in memory nor a previously
     returned pointer value should be used by a thread after calling any one
     of these three routines.  Memory allocated by a thread is automatically
     released on subsequent calls by the same thread to any of these three
     routines, and when the thread exits.

Since the code calls `getpwent` just once in the `parse_usersfile_str` function and doesn't pass the pointer out of the function scope, I believe it's safe to use `getpwnam` in this context.

**Comment by reneeotten** [comment]
@nickgaya @fracai so what's the deal with this PR? You want to merge something, this needs more work, or ...?

**Comment by fracai** [comment]
I'll defer to @nickgaya, but based on the last post it sounds like it's ready.

**Comment by nickgaya** [comment]
Thinking about this further, I realize this patch needs to be reworked again.  From the `getpwent(3)` man page:

> On OS X, these routines [`getpwnam` etc.] are thread-safe and return a pointer to a thread-specific data structure.  The contents of this data structure are automatically released by subsequent calls to any of these routines on the same thread, or when the thread exits.  **These routines are therefore unsuitable for use in libraries or frameworks, from where they may overwrite the per-thread data that the calling application expects to find as a result of its own calls to these routines. Library and framework code should use the alternative reentrant variants detailed below.**

**Comment by reneeotten** [comment]
> So I think there's no getting around the need for an implementation based on `getpwnam_r` with the attendant memory-management complexity.

well... I don't know anything about this port nor about the things you just wrote ;) What do you want to do with this PR: merge or close until you have reworked things.

**Comment by nickgaya** [comment]
Well it shouldn't be merged in its current state, so close I guess?  Or I can convert it back to a draft until I have a proper fix (hopefully in the next few days).

**Comment by nickgaya** [comment]
I updated the patch file, but it seems like it should more properly be submitted to the upstream project.

I opened https://gitlab.com/oath-toolkit/oath-toolkit/-/issues/26, awaiting feedback from the maintainers.

**Comment by reneeotten** [comment]
@nickgaya @fracai please decide what to do with this PR - either finish it up so that it can be merged of close it and re-open once there is a working solution.

**Comment by chenrui333** [comment]
I think there is some ongoing discussion in [here](https://gitlab.com/oath-toolkit/oath-toolkit/-/issues/26), so this PR might be very much still in progress?

**Comment by nickgaya** [comment]
There's been some recent progress on fixing the build issue upstream, but it may take some time until that's actually merged & released.

However, given that the only major change in 2.6.7 ([changelog](https://www.nongnu.org/oath-toolkit/NEWS.html)) was adding support for variable substitution in the pam_oath configuration, I don't think we have an urgent need to update the port.

I'm going to close this PR and recommend waiting until the next version of oath-toolkit to update.
